linkerd/proxy/transport/src/listen.rs

```diff
@@ -12,339 +12,75 @@ use tokio_stream::wrappers::TcpListenerStream;
-    pub fn new(addr: ListenAddr, keepalive: Keepalive, get_addrs: A) -> Self {
-        Self {
-            addr,
-            keepalive,
-            get_addrs,
-        }
-    }
-
-    pub fn with_orig_dst_addrs<A2>(self, get_addrs: A2) -> BindTcp<A2> {
-        BindTcp {
-            addr: self.addr,
-            keepalive: self.keepalive,
-            get_addrs,
-        }
-    }
-
-    fn accept(&self, tcp: TcpStream) -> io::Result<(A::Addrs, TcpStream)>
-    where
-        A: GetAddrs<TcpStream>,
-    {
-        let Keepalive(keepalive) = self.keepalive;
-        super::set_nodelay_or_warn(&tcp);
-        super::set_keepalive_or_warn(&tcp, keepalive);
-        let addrs = self.get_addrs.addrs(&tcp)?;
-
-        Ok((addrs, tcp))
+impl BindTcp {
+    pub fn with_orig_dst() -> super::BindWithOrigDst<Self> {
+        super::BindWithOrigDst::from(Self::default())
     }
 }
 
-impl<A> Bind for BindTcp<A>
+impl<T> Bind<T> for BindTcp
 where
-    A: GetAddrs<TcpStream> + Send + Sync + 'static,
-    A::Addrs: Send + Sync + 'static,
+    T: Param<ListenAddr> + Param<Keepalive>,
 {
-    type Addrs = A::Addrs;
+    type Addrs = Addrs;
     type Incoming = Pin<Box<dyn Stream<Item = io::Result<(Self::Addrs, Self::Io)>> + Send + Sync>>;
     type Io = TcpStream;
 
-    fn addr(&self) -> ListenAddr {
-        self.addr
-    }
-
-    fn keepalive(&self) -> Keepalive {
-        self.keepalive
-    }
-
-    fn bind(&self) -> io::Result<(SocketAddr, Self::Incoming)> {
+    fn bind(self, params: &T) -> io::Result<Bound<Self::Incoming>> {
         let listen = {
-            let l = std::net::TcpListener::bind(self.addr)?;
+            let ListenAddr(addr) = params.param();
+            let l = std::net::TcpListener::bind(addr)?;
             // Ensure that O_NONBLOCK is set on the socket before using it with Tokio.
             l.set_nonblocking(true)?;
             tokio::net::TcpListener::from_std(l).expect("listener must be valid")
         };
-        let addr = listen.local_addr()?;
-
-        let bind = self.clone();
-        let accept = TcpListenerStream::new(listen).map(move |tcp| bind.accept(tcp?));
-
-        Ok((addr, Box::pin(accept)))
-    }
-}
-
-impl Addrs {
-    pub fn new(
-        server: Local<ServerAddr>,
-        client: Remote<ClientAddr>,
-        orig_dst: OrigDstAddr,
-    ) -> Self {
-        Self {
-            server,
-            client,
-            orig_dst,
-        }
-    }
-
-    pub fn server(&self) -> Local<ServerAddr> {
-        self.server
-    }
-
-    pub fn client(&self) -> Remote<ClientAddr> {
-        self.client
-    }
+        let server = Local(ServerAddr(listen.local_addr()?));
+        let Keepalive(keepalive) = params.param();
+        let accept = TcpListenerStream::new(listen).map(move |res| {
+            let tcp = res?;
+            super::set_nodelay_or_warn(&tcp);
+            super::set_keepalive_or_warn(&tcp, keepalive);
+            let client = Remote(ClientAddr(tcp.peer_addr()?));
```

Comment:
small nitpick: I put the "accepted" trace event here once we have the addrs, and we don't have that any more...nbd but it might be worth having